src/main/kotlin/net/maizegenetics/phgv2/utils/GFFUtils.kt

```diff
@@ -0,0 +1,935 @@
+
+            // Create the Gff3Feature entry, add to list
+            val gff3FeatureEntry = Gff3FeatureImpl(entry.contig, entry.source, entry.type, newRange.first, newRange.last, entry.score, entry.strand, entry.phase, annotations)
+            pseudoGenomeGff3Features.add(gff3FeatureEntry)
+        }
+        // set offset to be offset plus length of all regions in the regionList
+        offset += regionList.sumOf { it.last - it.first + 1 } // add 1 because are inclusive/inclusive, we need a correct count
+
+    }
+
+    // At this point we need to determine if entries need to merge.  Or should I do that
+    // after the last one?
+    // Create the last "chromosome" entry for this file
+    val gff3FeatureEntry = createGffChromosomeEntry(curChrom, offset)
+    pseudoGenomeGff3Features.add(gff3FeatureEntry)
+
+    // Because we processed on a refRange basis, some entries could be overlapping
+    // reference ranges and thus appear twice.  these entries need to be merged. The
+    // following method does that.
+
+    val mergedFeatures = mergeAdjacentFeatures(pseudoGenomeGff3Features)
+
+    if (count > 0) total += count
+    myLogger.info("makeGffFromPath: all entries finished, total: ${total}")
+
+    if (outputFile != null && outputFile !="") {
+        myLogger.info("makeGffFromPath: writing pseudo gff file to ${outputFile}")
+        writeGffFile(outputFile, mergedFeatures.toSet(),null,null)
+    }
+    val endingTime = (System.nanoTime() - time)/1e9
+    myLogger.info("makeGffFromPath: time to process path: ${endingTime}")
+    return mergedFeatures.toSet()
+}
+
+/**
+ * when processing reference range at a time, some features may overlap
+ * multiple ranges.  The GFF file has no concept of reference ranges, so
+ * these entries need to be merged.
+ */
+fun mergeAdjacentFeatures(features: MutableSet<Gff3Feature>): List<Gff3Feature> {
+    // Sort features by contig, type, start, and end for easy processing
+    val sortedFeatures = features.sortedWith(compareBy({ it.contig }, { it.type }, { it.start }, { it.end }))
+    val mergedFeatures = mutableListOf<Gff3Feature>()
+
+    var idx = 0
+    while (idx < sortedFeatures.size) {
+        var currentFeature = sortedFeatures[idx]
+        var jdx = idx + 1
+        var hasMerged = false
+
+        // Check if the next feature is on the same chrom, is the same type and is adjacent or overlaps the current feature
+        while (jdx < sortedFeatures.size && featuresOverlap(currentFeature, sortedFeatures[jdx])
+        ) {
+
+            val nextFeature = sortedFeatures[jdx]
+
+            // Get referenceRangeID and halotypeAsmCoordinates attributes, handle nulls and lists
+            val currentRefRangeID = currentFeature.getAttribute("referenceRangeID")?.firstOrNull() ?: ""
+            val nextRefRangeID = nextFeature.getAttribute("referenceRangeID")?.firstOrNull() ?: ""
+            val newReferenceRangeID = listOf(currentRefRangeID, nextRefRangeID).joinToString(",")
```

Comment:
Why not just do this:
`val newReferenceRangeID = "${currentRefRangeID},${nextRefRangeID}"`

This will get weird if either are null, but the current way will also have the same issue.

Reply:
Changed